src/containers/Application/ApplicationPage.tsx

```diff
@@ -2,32 +2,96 @@ import React, { useEffect, useState } from 'react'
 }
+// TODO move to shared types if used elsewhere
+type SectionAndPage = { sectionCode: string; pageName: string } | null
+
+interface MethodToCallOnRevalidation {
+  (firstInvalidPage: SectionAndPage): void
+}
+
+const getFirstInvalidPage: (
+  fullStructure: FullStructure | undefined
+) => { sectionCode: string; pageName: string } | null = (fullStructure) => {
+  // return { sectionCode: 'S1', pageName: 'Page 2' }
+  return null
+}
+
+interface RevalidationState {
+  methodToCallOnRevalidation: MethodToCallOnRevalidation | null
+  shouldProcessValidation: boolean
+  lastRevalidationRequest: number
+}
 
 const ApplicationPage: React.FC<ApplicationProps> = ({ structure }) => {
-  const [strictSectionPage, setStrictSectionPage] = useState({ section: null, page: null })
-  const { error, isLoading, fullStructure, responsesByCode } = useGetFullApplicationStructure({
-    structure,
-  })
   const {
     userState: { currentUser },
   } = useUserState()
   const { push, query } = useRouter()
 
+  const {
+    state: { isLastElementUpdateProcessed, elementUpdatedTimestamp },
+  } = useFormElementUpdateTracker()
+
+  const [strictSectionPage, setStrictSectionPage] = useState<SectionAndPage>(null)
+  const [revalidationState, setRevalidationState] = useState<RevalidationState>({
+    methodToCallOnRevalidation: null,
+    shouldProcessValidation: false,
+    lastRevalidationRequest: Date.now(),
+  })
+
+  const shouldRevalidate = isLastElementUpdateProcessed && revalidationState.shouldProcessValidation
+  const revalidateAfterTimestamp = shouldRevalidate ? elementUpdatedTimestamp : 0
+
+  const { error, fullStructure } = useGetFullApplicationStructure({
+    structure,
+    shouldRevalidate,
+    revalidateAfterTimestamp,
+  })
+
   const currentSection = query.sectionCode
-  const currentPage = `Page ${query.page}`
+  const currentPage = `Page${query.page}`
+
+  /* Method to pass to progress bar, next button and submit button  to cause revalidation before aciton can be proceeded
+     Should always be called on submit, but only be called on next or progress bar navigation when isLinear */
+  // TODO may rename if we want to display loading modal ?
+  const requestRevalidation = (methodToCall: MethodToCallOnRevalidation) => {
+    setRevalidationState({
+      methodToCallOnRevalidation: methodToCall,
+      shouldProcessValidation: true,
+      lastRevalidationRequest: Date.now(),
+    })
+    // TODO show loading modal ?
+  }
+
+  // Revalidation Effect
+  useEffect(() => {
+    if (
+      fullStructure &&
+      revalidationState.methodToCallOnRevalidation &&
+      fullStructure.lastValidationTimestamp > revalidationState.lastRevalidationRequest
+    ) {
+      const lastValidPage = getFirstInvalidPage(fullStructure)
+      setRevalidationState({
+        ...revalidationState,
+        methodToCallOnRevalidation: null,
+        shouldProcessValidation: false,
+      })
+      if (lastValidPage != null) setStrictSectionPage(lastValidPage)
```

Comment:
```suggestion
      setStrictSectionPage(firstInvalidPage)
```

Reply:
Which should be `undefined` in case there isn't one invalidPage...

Reply:
One question: When running this for ProgressBar. For example you are currently on Page 2 and Page 3 is invalid, but if we click on a previousPage, then this method gets called and will set `firstInvalid` to Page 3, but we don't want to display any problems because current is Page 2 and the user clicked on Page 1. So maybe we need to receive here what is the page we are trying to do to?

Reply:
I had a few thought about this, it's probably not a good idea to set strict section page here automatically, I think it should be done by individual subscriber.

I think it might be best to pass through a method to set the strict section page to the subscriber:

how does that look ? 

5968da63adbdd1c34ea13e24dd73e85b928b744a

I.e. progress bar would get requestRevalidation through props, and use it on page/section navigation:
```
navigateToPage( () => {

  requestRevalidation( (firstInvalidPage, setStrictPage) => {

   // if firstInvalidPage === null, just navigate to that page
   // if firstInvalidPage is above or equal to page navigating to, just navigate to that page
   //  if firstInvalidPage is below page being navigated to, navigate to firstInvalidPage and setStrictPage(firstInvalidPage)
  })

})
```